submissions/trainperspective/trainperspective.md

```diff
@@ -0,0 +1,55 @@
+Step one would be to extract the raw image data (not to be confused with the .raw format) directly from factorio instead of encoding it in a png. I couldn’t find an easy way to do it, but turns out a bmp is very close. It’s the image data backwards with a header slapped in front. So this should be way faster to encode than a png.
+
+Next to get this into ffmpeg without saving to disk first. Turns out ffmpeg has built- in features for piping images, so a named pipe with a “.bmp” extension did the trick.
+
+Example command: ffmpeg -f image2pipe -framerate 60 -i - -r 60 -c:v libx264 -vf format=yuv420p -crf 1 example.mp4 -y < pipe.bmp
+Don’t forget to keep the pipe open between screenshots with a sleep command:
+sleep 10000000 > pipe.bmp, kill the sleep command to let ffmpeg finish the recording.
+
+
+I did a test run at a low fps and mhhhh, that’s not right… What’s going on here?
+
+![](render_420_2.mp4) (If tried to get it smaller, but if it’s still too big I put it on youtube too: https://youtu.be/jq4rL2V5Nss )
+
+The problem is that images are getting mixed. Factorio rendering is multithreaded. While one frame is still being written to the pipe, the next one might already start. Essentially mixing the pixels of both frames together.
+The fix is to force factorio to wait until the previous frame is rendered before starting the next one. This can be done with: “game.set_wait_for_screenshots_to_finish()” every frame.
+
+This however slows down factorio enough that we no longer can call it real time. Although I still have some ideas to speed it up, at this point I spend way too much time on this project already and decided to go with the tried and true method of using replays.
+Factorio has this wonderful feature, it doesn’t check if the content of the mods used during the recording and playback of a replay are the same.
+So first record a replay normally at a normal speed. Then edit one of your mods to take screenshots every tick. Then when you watch the replay factorio will start taking screenshots. That is assuming the changes you made to the mod don’t change the game state.
+
+Sadly due to using replays I had to cut a few scenes. I was going to show how weird it feels to build things from a rotated perspective. But it turns out factorio doesn’t save your mouse position in the replay. So in the replay the UI follows your mouse at the time of playback, not the position your mouse was when you recorded it.
```

Comment:
Is there any kind of video snippet of how the placing of buildings looked? That sounds super interesting.

Reply:
Don't think I still have these, I'll see if I can rerecord it.

Reply:
I have added some 2 video files from the exact same perspective, one captured directly the other one via replays. I haven't added them to the markdown file, maybe we can do side by side like you can do for images... is that possible?
I tried to make the video files as small as I could. But If you need to reprocess them (eg by uploading to a video site); I still have higher quality versions

Reply:
@Conor12345 is the one that usually compresses the videos using FFMPEG, so I'm not sure what his standards are. You can give us the lower res videos, but keep the higher res ones. We can always do a hot swop.

As for the side-by-side, I doubt this is possible for video, but I'll ask our website guys,

Reply:
@arrow-in-my-gluteus-maximus Please put the compressed and originals into the media folder for now and during the tech edit I'll decide on which will be best. Might be that I apply light compression to the high res originals to get decent quality whilst not being too big.

As for side-by-side/compare for video, @DrMagPie would be the best person to ask about that as I have no idea if its technically possible or not.

Reply:
I just added them to the same folder I added the LQ videos as I didn't find a media folder, and if you meant the assets folder it seems that one is split up by issue, so can't really be filled up until it's done. The side by side videos aren't actually higher resolution because it's just zoomed in, but it's higher bitrate. 

I also added hq versions of the other video files. I guess I was just hesitant to add too many large files cause it would clutter up the repo (don't want everyone to have to download a bunch of unused assets each time they clone the repo); but I guess now that you have repo access to my fork you can remove them there and then do a squash merge to avoid that.
Also it seems that DrMagPie has delivered

Reply:
I just added them to the same folder I added the LQ videos as I didn't find a media folder, and if you meant the assets folder it seems that one is split up by issue, so can't really be filled up until it's done. The side by side videos aren't actually higher resolution because it's just zoomed in, but it's higher bitrate.

I also added hq versions of the other video files. I guess I was just hesitant to add too many large files cause it would clutter up the repo (don't want everyone to have to download a bunch of unused assets each time they clone the repo); but I guess now that you have repo access to my fork you can remove them there and then do a squash merge to avoid that.
Also it seems that DrMagPie has delivered

Reply:
Media is not stored in the repo. So when this PR is merged Conor will move remove it from the repo, and move it to our media server.

But looks like we will have side-by-side videos ready for next week. So it would be awesome if you could add it to this file. Just specifying which video to place left/right, and what caption to add.

Reply:
added it